presto-main/src/test/java/com/facebook/presto/sql/gen/TestCursorProcessorCompiler.java

```diff
@@ -0,0 +1,221 @@
+            LESS_THAN.name(),
+            FUNCTION_MANAGER.resolveOperator(LESS_THAN, fromTypes(BIGINT, BIGINT)),
+            BOOLEAN,
+            ADD_X_Y,
+            constant(10L, BIGINT));
+
+    private static final CallExpression ADD_X_Y_Z = call(
+            ADD.name(),
+            FUNCTION_MANAGER.resolveOperator(ADD, fromTypes(BIGINT, BIGINT)),
+            BIGINT,
+            call(
+                    ADD.name(),
+                    FUNCTION_MANAGER.resolveOperator(ADD, fromTypes(BIGINT, BIGINT)),
+                    BIGINT,
+                    field(0, BIGINT),
+                    field(1, BIGINT)),
+            field(2, BIGINT));
+
+    @Test
+    public void testRewriteRowExpressionWithCSE()
+    {
+        RowExpression filter = new SpecialFormExpression(AND, BIGINT, ADD_X_Y_GREATER_THAN_2);
+        ImmutableList<RowExpression> rowExpressions = ImmutableList.of(ADD_X_Y_Z, filter);
+        Map<LambdaDefinitionExpression, LambdaBytecodeGenerator.CompiledLambda> compiledLambdaMap = generateMethodsForLambda(cursorProcessorClassDefinition, callSiteBinder, cachedInstanceBinder, rowExpressions, meta, sqlFunctionProperties, "");
+
+        Map<Integer, Map<RowExpression, VariableReferenceExpression>> commonSubExpressionsByLevel = collectCSEByLevel(rowExpressions);
+
+        Map<VariableReferenceExpression, CommonSubExpressionRewriter.CommonSubExpressionFields> cseFields = declareCommonSubExpressionFields(cursorProcessorClassDefinition, commonSubExpressionsByLevel);
+        Map<RowExpression, VariableReferenceExpression> commonSubExpressions = commonSubExpressionsByLevel.values().stream()
+                .flatMap(m -> m.entrySet().stream())
+                .collect(toImmutableMap(Map.Entry::getKey, Map.Entry::getValue));
+        // X+Y as CSE
+        assertEquals(1, cseFields.size());
+        VariableReferenceExpression cseVariable = cseFields.keySet().iterator().next();
+
+        RowExpression rewrittenFilter = cseCursorCompiler.rewriteRowExpressionsWithCSE(ImmutableList.of(filter), commonSubExpressions).get(0);
+
+        List<RowExpression> rewrittenProjections = cseCursorCompiler.rewriteRowExpressionsWithCSE(ImmutableList.of(ADD_X_Y_Z), commonSubExpressions);
+
+        // X+Y+Z contains CSE X+Y
+        assertTrue(((CallExpression) rewrittenProjections.get(0)).getArguments().contains(cseVariable));
+
+        // X+Y > 2 consists CSE X+Y
+        assertTrue(((CallExpression) ((SpecialFormExpression) rewrittenFilter).getArguments().get(0)).getArguments().contains(cseVariable));
+    }
+
+    @Test
+    public void testCompilerWithCSE()
+    {
+        PageFunctionCompiler functionCompiler = new PageFunctionCompiler(meta, 0);
+        ExpressionCompiler expressionCompiler = new ExpressionCompiler(meta, functionCompiler);
+
+        RowExpression filter = new SpecialFormExpression(AND, BIGINT, ADD_X_Y_GREATER_THAN_2, ADD_X_Y_LESS_THAN_10);
+        List<? extends RowExpression> projections = createIfProjectionList(5);
```

Comment:
Would be interesting to make the filter and projection share some cse.

Reply:
@rongrong mk sense. add `ADD_X_Y` to the projections so "X+Y" is shared between projections & filter